server/service/apple_mdm.go

```diff
@@ -3438,6 +3471,249 @@ func ReconcileAppleProfiles(
 	return nil
 }
 
+func preprocessProfileContents(
+	ctx context.Context,
+	appConfig *fleet.AppConfig,
+	ds fleet.Datastore,
+	targets map[string]*cmdTarget,
+	profileContents map[string]mobileconfig.Mobileconfig,
+) error {
+
+	// This method replaces Fleet variables ($FLEET_VAR_<NAME>) in the profile contents, generating a unique profile for each host.
+	// For a 2KB profile and 30K hosts, this method may generate ~60MB of profile data in memory.
+
+	isNDESSCEPConfigured := func(profUUID string, target *cmdTarget) (bool, error) {
+		if !license.IsPremium(ctx) {
+			for _, hostUUID := range target.hostUUIDs {
+				err := ds.UpdateOrDeleteHostMDMAppleProfile(ctx, &fleet.HostMDMAppleProfile{
+					CommandUUID:   target.cmdUUID,
+					HostUUID:      hostUUID,
+					Status:        &fleet.MDMDeliveryFailed,
+					Detail:        "NDES SCEP Proxy requires a Fleet Premium license.",
+					OperationType: fleet.MDMOperationTypeInstall,
+				})
+				if err != nil {
+					return false, err
+				}
+			}
+			return false, nil
+		}
+		if !appConfig.Integrations.NDESSCEPProxy.Valid {
+			for _, hostUUID := range target.hostUUIDs {
+				err := ds.UpdateOrDeleteHostMDMAppleProfile(ctx, &fleet.HostMDMAppleProfile{
+					CommandUUID: target.cmdUUID,
+					HostUUID:    hostUUID,
+					Status:      &fleet.MDMDeliveryFailed,
+					Detail: "NDES SCEP Proxy is not configured. " +
+						"Please configure in Settings > Integrations > Mobile Device Management > Simple Certificate Enrollment Protocol.",
+					OperationType: fleet.MDMOperationTypeInstall,
+				})
+				if err != nil {
+					return false, err
+				}
+			}
+		}
+		return appConfig.Integrations.NDESSCEPProxy.Valid, nil
+	}
+
+	// Copy of NDES SCEP config which will contain unencrypted password, if needed
+	var ndesConfig *fleet.NDESSCEPProxyIntegration
+
+	var addedTargets map[string]*cmdTarget
+	for profUUID, target := range targets {
+		contents, ok := profileContents[profUUID]
+		if !ok {
+			// This should never happen
+			continue
+		}
+
+		// Check if Fleet variables are present.
+		contentsStr := string(contents)
+		fleetVars := findFleetVariables(contentsStr)
+		if len(fleetVars) == 0 {
+			continue
+		}
+
+		// Do common validation that applies to all hosts in the target
+		valid := true
+		for fleetVar := range fleetVars {
+			switch fleetVar {
+			case FleetVarNDESSCEPChallenge, FleetVarNDESSCEPProxyURL:
+				configured, err := isNDESSCEPConfigured(profUUID, target)
+				if err != nil {
+					return ctxerr.Wrap(ctx, err, "checking NDES SCEP configuration")
+				}
+				if !configured {
+					valid = false
+					break
+				}
+			case FleetVarHostEndUserEmailIDP:
+				// No extra validation needed for this variable
+			default:
+				// Error out if we find an unknown variable
```

Comment:
I guess this is coming from product, but isn't it weird we error here and not when the profile is uploaded?

Reply:
Yes, that makes sense. We should validate known variables on upload. I will add that check to the next PR.